diff --git a/mod_wsgi.c b/mod_wsgi.c
index f9ffe09..722261c 100644
--- a/mod_wsgi.c
+++ b/mod_wsgi.c
@@ -10696,6 +10696,19 @@ static void wsgi_daemon_main(apr_pool_t *p, WSGIDaemonProcess *daemon)
     poll_fd.reqevents = APR_POLLIN;
     poll_fd.desc.f = wsgi_signal_pipe_in;
 
+    /*
+    * Register signal handler to receive shutdown signal
+    * from Apache parent process.
+    */
+
+    wsgi_daemon_shutdown = 0;
+
+    apr_signal(SIGINT, wsgi_signal_handler);
+    apr_signal(SIGTERM, wsgi_signal_handler);
+#ifdef SIGXCPU
+    apr_signal(SIGXCPU, wsgi_signal_handler);
+#endif
+
     /* Initialise maximum request count for daemon. */
 
     if (daemon->group->maximum_requests)
@@ -11082,19 +11095,6 @@ static int wsgi_start_process(apr_pool_t *p, WSGIDaemonProcess *daemon)
             }
         }
 
-        /*
-         * Register signal handler to receive shutdown signal
-         * from Apache parent process.
-         */
-
-        wsgi_daemon_shutdown = 0;
-
-        apr_signal(SIGINT, wsgi_signal_handler);
-        apr_signal(SIGTERM, wsgi_signal_handler);
-#ifdef SIGXCPU
-        apr_signal(SIGXCPU, wsgi_signal_handler);
-#endif
-
         /* Set limits on amount of CPU time that can be used. */
 
         if (daemon->group->cpu_time_limit > 0) {
